File: uva401.c
#include <ctype.h>
#include <stdio.h>
#include <string.h>

const char *rev = "A   3  HIL JM O   2TUVWXY51SE Z  8 ";
const char *msg[] = {"not a palindrome", "a regular palindrome",
                     "a mirrored string", "a mirrored palindrome"};

char r(char ch) {
  if (isalpha(ch))
    return rev[ch - 'A'];
  return rev[ch - '0' + 25];
}

int main(void) {
  char s[30];
  while (scanf("%s", s) == 1) {
    int len = strlen(s);
    int p = 1, m = 1;

    int i;
    for (i = 0; i < (len + 1) / 2; i++) {
      /* 不是回文 */
      if (s[i] != s[len - i - 1])
        p = 0;

      /* 不是镜像 */
      if (r(s[i]) != s[len - i - 1])
        m = 0;
    }
    printf("%s -- is %s.\n\n", s, msg[m * 2 + p]);
  }

  return 0;
}

File: src/b1ch03/uva401.java
package b1ch03;

import java.util.Scanner;

public class uva401 {

    private static char[] rev = "A   3  HIL JM O   2TUVWXY51SE Z  8 ".toCharArray();
    private static String[] description = {
            "not a palindrome", "a regular palindrome",
            "a mirrored string", "a mirrored palindrome"
    };

    private static char r(char ch) {
        if (ch >= 'A' && ch <= 'Z') return rev[ch - 'A'];
        return rev[ch - '0' + 25];
    }

    private static int judge(char[] array) {
        int p = 1, m = 1;
        for (int i = 0; i < array.length; i++) {
            if (array[i] != array[array.length - 1 - i]) p = 0;
            if (r(array[i]) != array[array.length - 1 - i]) m = 0;
        }
        return m * 2 + p;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        while (sc.hasNext()) {
            String str = sc.nextLine();
            System.out.print(
                    String.format("%s -- is %s.\n\n", str,
                            description[judge(str.toCharArray())])
            );
        }
    }

}
